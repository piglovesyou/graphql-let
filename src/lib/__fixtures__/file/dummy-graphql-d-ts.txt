/* a7870383395320c75dc806f677ffa7caad7522bc
 * This file is automatically generated by graphql-let.
 */
