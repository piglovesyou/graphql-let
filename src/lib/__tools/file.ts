import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
