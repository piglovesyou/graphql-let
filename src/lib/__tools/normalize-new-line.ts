const crlf = /\r\n/g;

export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}
