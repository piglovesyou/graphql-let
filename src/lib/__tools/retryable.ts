function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default async function retryable(
  fn: () => Promise<void>,
  interval: number,
  maxTimeout: number,
) {
  if (maxTimeout < 0) throw new Error('never');
  const end = Date.now() + maxTimeout;
  let lastError: Error = new Error('never');
  let n = 0;
  while (Date.now() < end) {
    try {
      await fn();
      console.info(`Succeeded on the ${++n} time.`);
      return;
    } catch (e) {
      console.info(`${++n} times tried. Retrying...`);
      lastError = e;
      await timeout(interval);
    }
  }
  throw lastError;
}
