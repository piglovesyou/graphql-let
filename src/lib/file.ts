import { promises as fsPromises } from 'fs';
import { createReadStream, existsSync } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;
export { readFileSync, statSync } from 'fs';

// Erasing old cache in __generated__ on HMR.
// Otherwise the multiple `declare module "*/x.graphql"` are exposed.
export async function removeByPatterns(cwd: string, ...patterns: string[]) {
  const oldFiles = await glob(patterns, {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map((f) => rimraf(f)));
}

const leadingStringOfGeneratedContent = '/* ';
const hexHashLength = 40;
export function withHash(sourceHash: string, content: string): string {
  return `${leadingStringOfGeneratedContent}${sourceHash}
 * This file is automatically generated by graphql-let. */

${content}`;
}

export function readHash(filePath: string): Promise<string | null> {
  if (!existsSync(filePath)) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const stream = createReadStream(filePath, {
      encoding: 'utf-8',
      highWaterMark: leadingStringOfGeneratedContent.length + hexHashLength,
    });
    stream.on('error', (error) => reject(error));
    stream.on('data', (chunk: string) => {
      const hash = chunk.slice(leadingStringOfGeneratedContent.length);
      if (hash.length !== hexHashLength) return resolve(null);

      resolve(hash);
    });
    stream.read();
  });
}
