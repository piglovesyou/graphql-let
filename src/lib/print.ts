const PREFIX = '[ graphql-let ]';

export function printInfo(message: string): void {
  console.info([PREFIX, message].join('\t'));
}

export function printError(err: Error): void {
  console.error([PREFIX, err.message].join(' '));
}
