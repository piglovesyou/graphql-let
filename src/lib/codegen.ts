import { codegen as graphqlCodegen } from '@graphql-codegen/core';
import { promises as fsPromises } from 'fs';
import gql from 'graphql-tag';
import _mkdirp from 'mkdirp';
import path from 'path';
import { promisify } from 'util';
import genDts from './gen-dts';
import { PartialCodegenOpts } from './create-codegen-opts';
import { printInfo } from './print';
import { ConfigTypes } from './types';
import { existsSync } from 'fs';

const { readFile, writeFile } = fsPromises;
const mkdirp = promisify(_mkdirp);

// For the loader can process a same file simultaneously
const processingTasks = new Map<string /*fileFullPath*/, Promise<string>>();

function wrapAsModule(fileName: string, content: string) {
  return `declare module '*/${fileName}' {
  ${content.replace(/\n/g, '\n  ')}}`;
}

async function processGraphQLCodegen(
  codegenOpts: PartialCodegenOpts,
  tsxFullPath: string,
  gqlFullPath: string,
  gqlContent: string,
): Promise<string> {
  const tsxContent = await graphqlCodegen({
    ...codegenOpts,
    filename: tsxFullPath,
    documents: [
      {
        filePath: gqlFullPath,
        content: gql(gqlContent),
      },
    ],
  });
  await mkdirp(path.dirname(tsxFullPath));
  await writeFile(tsxFullPath, tsxContent);
  return tsxContent;
}

async function processGenDts(
  dtsFullPath: string,
  tsxFullPath: string,
  gqlFullPath: string,
  dtsRelPath: string,
) {
  await mkdirp(path.dirname(dtsFullPath));
  const dtsContent = await genDts(tsxFullPath);
  if (!dtsContent) throw new Error(`Generate ${dtsFullPath} fails.`);
  await writeFile(
    dtsFullPath,
    wrapAsModule(path.basename(gqlFullPath), dtsContent),
  );
  printInfo(`${dtsRelPath} was generated.`);
  return dtsContent;
}

export async function codegen(
  gqlContent: string,
  gqlFullPath: string,
  tsxFullPath: string,
  dtsRelPath: string,
  dtsFullPath: string,
  options: ConfigTypes,
  codegenOpts: PartialCodegenOpts,
): Promise<string> {
  let tsxContent: string;
  if (existsSync(tsxFullPath)) {
    tsxContent = await readFile(tsxFullPath, 'utf-8');
  } else if (processingTasks.has(tsxFullPath)) {
    tsxContent = await processingTasks.get(tsxFullPath)!;
  } else {
    const tsxPromise = processGraphQLCodegen(
      codegenOpts,
      tsxFullPath,
      gqlFullPath,
      gqlContent,
    );
    processingTasks.set(tsxFullPath, tsxPromise);
    tsxContent = await tsxPromise;
    processingTasks.delete(tsxFullPath);
  }

  if (existsSync(dtsFullPath) || processingTasks.has(dtsFullPath)) {
    // Already exists or is processing. Just skip.
  } else {
    const dtsPromise = processGenDts(
      dtsFullPath,
      tsxFullPath,
      gqlFullPath,
      dtsRelPath,
    );
    processingTasks.set(dtsFullPath, dtsPromise);
    await dtsPromise;
    processingTasks.delete(dtsFullPath);
  }

  return tsxContent;
}
