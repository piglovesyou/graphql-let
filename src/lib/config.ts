import { Types } from '@graphql-codegen/plugin-helpers';
import { join as pathJoin } from 'path';
import { parse as parseYaml } from 'yaml';
import { DEFAULT_CONFIG_FILENAME } from './consts';
import { readFile, readFileSync } from './file';
import { createHash } from './hash';
import { printError } from './print';

export type PartialGraphqlCodegenOptions = Omit<Types.Config, 'generates'>;

export type GraphQLLetAdditionalOptions = {
  plugins: Array<string | Record<string, any>>;
  respectGitIgnore?: boolean;
  schemaEntrypoint?: string;
  cacheDir?: string;
  TSConfigFile?: string;
  gqlDtsEntrypoint?: string;
  generateOptions?: Types.ConfiguredOutput;
};

export type UserConfigTypes = PartialGraphqlCodegenOptions &
  GraphQLLetAdditionalOptions;
export type ConfigTypes = PartialGraphqlCodegenOptions & {
  documents: Types.OperationDocumentGlobPath[];
} & Required<GraphQLLetAdditionalOptions>;

export function buildConfig(raw: UserConfigTypes): ConfigTypes {
  if (typeof raw !== 'object')
    printError(new Error('A config file must shape an object'));

  if (!raw.schema || !raw.documents || !raw.plugins)
    printError(new Error(`A config requires a "${name}" field`));

  const documents: Types.OperationDocumentGlobPath[] = Array.isArray(
    raw.documents,
  )
    ? (raw.documents as Types.OperationDocumentGlobPath[])
    : typeof raw.documents === 'string'
    ? [raw.documents]
    : (printError(
        new Error(`config.documents should be an array or a string`),
      ) as never);

  return {
    ...raw,
    // Normalized codegen options
    documents,
    // Set graphql-let default values
    respectGitIgnore: raw.respectGitIgnore || true,
    cacheDir: raw.cacheDir || 'node_modules/graphql-let/__generated__',
    TSConfigFile: raw.TSConfigFile || 'tsconfig.json',
    gqlDtsEntrypoint:
      raw.gqlDtsEntrypoint || 'node_modules/@types/graphql-let/index.d.ts',
    generateOptions: raw.generateOptions || Object.create(null),
    schemaEntrypoint: raw.schemaEntrypoint || '',
  };
}

export const getConfigPath = (cwd: string, configFilePath?: string) =>
  pathJoin(cwd, configFilePath || DEFAULT_CONFIG_FILENAME);

const getConfigFromContent = (content: string): [ConfigTypes, string] => [
  buildConfig(parseYaml(content)),
  createHash(content),
];

export default async function loadConfig(
  cwd: string,
  configFilePath?: string,
): Promise<[ConfigTypes, string]> {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = await readFile(configPath, 'utf-8');
  return getConfigFromContent(content);
}

export function loadConfigSync(
  cwd: string,
  configFilePath?: string,
): [ConfigTypes, string] {
  const configPath = getConfigPath(cwd, configFilePath);
  const content = readFileSync(configPath, 'utf-8');
  return getConfigFromContent(content);
}
