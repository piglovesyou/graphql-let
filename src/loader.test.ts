/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { ok, strictEqual } from 'assert';
import 'core-js/es/array';
import * as fs from 'fs';
import glob from 'globby';
import { join as pathJoin } from 'path';
import { promisify } from 'util';
import waitOn from 'wait-on';
import { Stats } from 'webpack';
import * as print from './lib/print';
import compiler from './lib/__tools/compile';
import { prepareFixtures } from './lib/__tools/file';

const unlink = promisify(fs.unlink);

let basicFixtureDir: string;
let monorepoFixtureDir: string;
let silentFixtureDir: string;

const basicEntrypoints = ['pages/index.tsx', 'pages/viewer.graphql'];

function getOutputInfo(stats: Stats) {
  const { modules } = stats.toJson()!;
  const acc: [name: string, output: string][] = [];
  (function getNamesAndOutputs(fnModules: Stats.FnModules[]) {
    for (const m of fnModules) {
      if (m.name && m.source) acc.push([m.name, m.source]);
      if (m.modules) getNamesAndOutputs(m.modules);
    }
  })(modules!);
  return acc;
}

describe('graphql-let/loader', () => {
  beforeAll(async () => {
    [basicFixtureDir] = await prepareFixtures(
      __dirname,
      '__fixtures/loader/basic',
    );
    [monorepoFixtureDir] = await prepareFixtures(
      __dirname,
      '__fixtures/loader/monorepo',
    );
    [silentFixtureDir] = await prepareFixtures(
      __dirname,
      '__fixtures/loader/silent',
    );
  });

  test('generates .tsx and .d.ts', async () => {
    const stats = await compiler(basicFixtureDir, basicEntrypoints, 'node');
    const outputs = getOutputInfo(stats);
    expect(outputs).toHaveLength(4);

    const [, gqlOutput] = outputs.find(
      ([name]) => name === './pages/viewer.graphql',
    )!;
    expect(gqlOutput).toMatchSnapshot();
    const [, sourceOutput] = outputs.find(
      ([name]) => name === './pages/index.tsx',
    )!;
    expect(sourceOutput).toMatchSnapshot();
  });

  test('runs well for simultaneous execution, assuming SSR', async () => {
    const expectedTargets: [string, 'node' | 'web'][] = [
      ['pages/index.tsx', 'node'],
      ['pages/index.tsx', 'web'],
      ['pages/viewer.graphql', 'node'],
      ['pages/viewer.graphql', 'web'],
      ['pages/viewer2.graphql', 'node'],
      ['pages/viewer2.graphql', 'web'],
    ];
    const results = await Promise.all(
      expectedTargets.map(([file, target]) =>
        compiler(basicFixtureDir, [file], target),
      ),
    );
    for (const [i, stats] of results.entries()) {
      const [file] = expectedTargets[i];
      const outputs = getOutputInfo(stats);
      const [, output] = outputs.find(([name]) => name === './' + file)!;
      expect(output).toMatchSnapshot();
    }
    const globResults = await glob('**/*.graphql.d.ts', {
      cwd: basicFixtureDir,
    });
    strictEqual(globResults.length, 2);
  });

  test('The option "silent" suppresses standard output logs', async () => {
    let messages = '';
    const mockFn = (m: any) => (messages += m + '\n');
    jest.spyOn(print, 'printInfo').mockImplementation(mockFn);
    jest.spyOn(print, 'updateLog').mockImplementation(mockFn);

    await compiler(basicFixtureDir, basicEntrypoints, 'node');
    expect(messages).toMatchSnapshot();

    messages = '';
    await compiler(silentFixtureDir, basicEntrypoints, 'node');
    expect(messages).toHaveLength(0);
  });

  describe('options', () => {
    async function acceptsConfigPathInOptionsConfigFile(
      configFilePath: string,
    ) {
      const stats = await compiler(
        pathJoin(monorepoFixtureDir, 'packages/app'),
        ['src/index.ts'],
        'web',
        { configFile: configFilePath },
      );

      const modules = stats
        .toJson()
        .modules?.flatMap((m) => m.modules)
        ?.filter(Boolean);

      ok(modules);
      expect(modules.map((m) => m.name)).toMatchInlineSnapshot(`
        Array [
          "./src/index.ts",
          "./src/fruits.graphql",
        ]
      `);

      const generated = modules.find((m) => m?.name === './src/fruits.graphql');

      ok(generated);

      await waitOn({
        resources: [
          `${monorepoFixtureDir}/packages/app/__generated__/src/fruits.graphql.tsx`,
        ],
      });

      expect(generated.source).toContain('export function useGetFruitsQuery');
      expect(
        generated.source?.replace(/\/\*[\s\S]*?\*\//g, ''),
      ).toMatchSnapshot();

      await Promise.all([
        unlink(
          `${monorepoFixtureDir}/packages/app/__generated__/src/fruits.graphql.tsx`,
        ),
        unlink(`${monorepoFixtureDir}/packages/app/src/fruits.graphql.d.ts`),
      ]).catch(() => {
        /* discard error */
      });
    }

    test('accept relative config path in options.configFile', async () => {
      await acceptsConfigPathInOptionsConfigFile(
        '../../config/.graphql-let.yml',
      );
    });

    test('accept absolute config path in options.configFile', async () => {
      await acceptsConfigPathInOptionsConfigFile(
        require.resolve('./__fixtures/loader/monorepo/config/.graphql-let.yml'),
      );
    });
  });
});
