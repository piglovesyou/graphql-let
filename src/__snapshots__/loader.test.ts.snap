// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-let/loader generates .tsx and .d.ts 1`] = `
"/* 65181ff8ca502f8dcdf828f03f90b7e048d3f6a9
 * This file is automatically generated by graphql-let. */
import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export const ViewerDocument = gql\`
    query Viewer {
  viewer {
    id
    name
    status
  }
}
    \`;
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

export function useViewerQuery(baseOptions) {
  return Apollo.useQuery(ViewerDocument, baseOptions);
}
export function useViewerLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(ViewerDocument, baseOptions);
}"
`;

exports[`graphql-let/loader options accept absolute config path in options.configFile 2`] = `
"import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export const GetFruitsDocument = gql\`
    query GetFruits {
  fruits {
    id
    name
    flavor
  }
}
    \`;


export function useGetFruitsQuery(baseOptions) {
  return Apollo.useQuery(GetFruitsDocument, baseOptions);
}
export function useGetFruitsLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(GetFruitsDocument, baseOptions);
}"
`;

exports[`graphql-let/loader options accept relative config path in options.configFile 2`] = `
"import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export const GetFruitsDocument = gql\`
    query GetFruits {
  fruits {
    id
    name
    flavor
  }
}
    \`;


export function useGetFruitsQuery(baseOptions) {
  return Apollo.useQuery(GetFruitsDocument, baseOptions);
}
export function useGetFruitsLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(GetFruitsDocument, baseOptions);
}"
`;
